### ISSUE TITLE
spotify's  window is not detected by spotify-tray on version 1.1.67.586

### ISSUE THREAD

**Issue by Moitaeel** [tsmetana/spotify-tray#22]
spotify-tray launches the spotify program as expected but doesn't recognize the window, so it closes. It works on downgraded version 1.1.55.498.

Also, the window has no icon on KDE/Plasma, either launched by spotify-tray or the program itself on terminal, so something weird is happening on spotify side.

![Screenshot_20210905_223042](https://user-images.githubusercontent.com/57826296/132148509-1cf8bd12-f55a-4d47-a586-f5c5b1c942cb.png)

Terminal output from **working** Spotify version 1.1.55.498

```
$ spotify-tray 
spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by spotify)
/opt/spotify/spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /opt/spotify/spotify)
/opt/spotify/spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /opt/spotify/spotify)
/proc/self/exe: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /proc/self/exe)
````

Terminal output from **problematic** Spotify version 1.1.67.586

```
$ spotify-tray 
spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by spotify)
/opt/spotify/spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /opt/spotify/spotify)
/opt/spotify/spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /opt/spotify/spotify)
/proc/self/exe: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /proc/self/exe)

** (spotify-tray:269665): WARNING **: 23:22:41.436: Could not find the Spotify client window: attempting to launch the application, attempt 1/5

** (spotify-tray:269665): WARNING **: 23:22:41.937: Could not find the Spotify client window: attempting to launch the application, attempt 2/5

** (spotify-tray:269665): WARNING **: 23:22:42.438: Could not find the Spotify client window: attempting to launch the application, attempt 3/5

** (spotify-tray:269665): WARNING **: 23:22:42.946: Could not find the Spotify client window: attempting to launch the application, attempt 4/5

** (spotify-tray:269665): WARNING **: 23:22:43.446: Could not find the Spotify client window: attempting to launch the application, attempt 5/5

** (spotify-tray:269665): CRITICAL **: 23:22:43.446: Could not find the Spotify client window: giving up
```

**Comment by tsmetana** [comment]
I'm unable to reproduce the problem with spotify client version 1.1.67.586.gbb5ef64e on Fedora 33. So I need to ask you to help me debugging this. The first thing to try would be to verify the `WM_CLASS` property of the Spotify client window since that's how the spotify-tray is identifying the client window (should be "spotify"): What is the output of `xprop | grep WM_CLASS`? You may also start spotify-tray with `G_MESSAGES_DEBUG=all ./spotify-tray`, it may spit few more messages, maybe it would give some hint.

**Comment by Moitaeel** [comment]
Hello @tsmetana

I forgot to add my system's info, I'm on Manjaro, testing repo.
<details>
<summary>System info</summary>

```
Operating System: Manjaro Linux
KDE Plasma Version: 5.22.5
KDE Frameworks Version: 5.85.0
Qt Version: 5.15.2
Kernel Version: 5.10.61-1-MANJARO (64-bit)
Graphics Platform: X11
Processors: 6 × Intel® Core™ i7-8750H CPU @ 2.20GHz
Memory: 15,3 GiB of RAM
Graphics Processor: Mesa Intel® UHD Graphics 630
Xorg Server Version: 1.20.13-3
````
</details>

OK, I tried to run `xprop | grep WM_CLASS` on spotify's window but there's no output, I tried clicking on anything else and it all works except spotify
<details>
<summary>xprop | grep WM_CLASS</summary>

```
$ xprop | grep WM_CLASS
WM_CLASS(STRING) = "konsole", "konsole"
$ xprop | grep WM_CLASS #clicking on spotify's window
$ xprop | grep WM_CLASS #clicking on spotify's window
$ xprop | grep WM_CLASS #clicking on spotify's window
$ xprop | grep WM_CLASS
WM_CLASS(STRING) = "plasmashell", "plasmashell"
$ xprop | grep WM_CLASS
WM_CLASS(STRING) = "ferdi", "Ferdi"
$ xprop | grep WM_CLASS
WM_CLASS(STRING) = "Conky", "Conky"
```
</details>

Running `G_MESSAGES_DEBUG=all ./spotify-tray` gave this output:
<details>
<summary>G_MESSAGES_DEBUG=all ./spotify-tray</summary>

```
$ G_MESSAGES_DEBUG=all ./spotify-tray
(spotify-tray:44974): GLib-GIO-DEBUG: 20:22:55.444: _g_io_module_get_default: Found default implementation gvfs (GDaemonVfs) for ‘gio-vfs’
** (spotify-tray:44974): DEBUG: 20:22:55.497: D-Bus method 'RaiseWindow' call failed: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name name.smetana.SpotifyTray was not provided by any .service files
spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by spotify)
/opt/spotify/spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /opt/spotify/spotify)
/opt/spotify/spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /opt/spotify/spotify)
(spotify:44977): GLib-GIO-DEBUG: 20:22:55.655: _g_io_module_get_default: Found default implementation gvfs (GDaemonVfs) for ‘gio-vfs’
/proc/self/exe: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /proc/self/exe)

** (spotify-tray:44974): WARNING **: 20:22:55.999: Could not find the Spotify client window: attempting to launch the application, attempt 1/5

** (spotify-tray:44974): WARNING **: 20:22:56.500: Could not find the Spotify client window: attempting to launch the application, attempt 2/5

** (spotify-tray:44974): WARNING **: 20:22:57.001: Could not find the Spotify client window: attempting to launch the application, attempt 3/5

** (spotify-tray:44974): WARNING **: 20:22:57.502: Could not find the Spotify client window: attempting to launch the application, attempt 4/5

** (spotify-tray:44974): WARNING **: 20:22:58.003: Could not find the Spotify client window: attempting to launch the application, attempt 5/5

** (spotify-tray:44974): CRITICAL **: 20:22:58.003: Could not find the Spotify client window: giving up
```

</details>

As for comparison, I ran the same commands on spotify **1.1.55.498** and these are the outputs:

<details>
<summary>xprop | grep WM_CLASS</summary>

```
$ xprop | grep WM_CLASS #clicking on spotify's window
WM_CLASS(STRING) = "spotify", "Spotify"
```

</details>

<details>
<summary>G_MESSAGES_DEBUG=all ./spotify-tray</summary>

```
$ G_MESSAGES_DEBUG=all ./spotify-tray
(spotify-tray:73577): GLib-GIO-DEBUG: 20:35:05.334: _g_io_module_get_default: Found default implementation gvfs (GDaemonVfs) for ‘gio-vfs’
** (spotify-tray:73577): DEBUG: 20:35:05.386: D-Bus method 'RaiseWindow' call failed: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name name.smetana.SpotifyTray was not provided by any .service files
spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by spotify)
/opt/spotify/spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /opt/spotify/spotify)
/opt/spotify/spotify: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /opt/spotify/spotify)
(spotify:73580): GLib-GIO-DEBUG: 20:35:05.536: _g_io_module_get_default: Found default implementation gvfs (GDaemonVfs) for ‘gio-vfs’
/proc/self/exe: /usr/lib/libcurl-gnutls.so.4: no version information available (required by /proc/self/exe)
** (spotify-tray:73577): DEBUG: 20:35:05.889: Class pid 73580
** (spotify-tray:73577): DEBUG: 20:35:05.956: Found icon: spotify

** (spotify-tray:73577): CRITICAL **: 20:35:51.833: Failed to retrieve the Metadata property

** (spotify-tray:73577): CRITICAL **: 20:35:52.333: The applicaiton has exited, quitting.
```

</details>

**Comment by tsmetana** [comment]
> OK, I tried to run `xprop | grep WM_CLASS` on spotify's window but there's no output, I tried clicking on anything else and it all works except spotify

This looks like the WM_CLASS is not set. I don't know how is your spotify client application installed but the fact that spotify-tray can't find neither the window nor the icon means that you would either have to live without the tray icon or try some workarounds for the missing system integration: set the WM_CLASS from the desktop file starting the app (not sure if that would work with new spotify that double forks, but worth trying) and spotify-tray can be told via command line option which icon file to use.

To verify that it's really the case, you may try to set the WM_CLASS to the spotify window with `xprop -f WM_CLASS 8s -set WM_CLASS "spotify"` (it has to be running) and after that start the spotify-tray: it should be able to find the running app and use it.

**Comment by Moitaeel** [comment]
> To verify that it's really the case, you may try to set the WM_CLASS to the spotify window with `xprop -f WM_CLASS 8s -set WM_CLASS "spotify"` (it has to be running) and after that start the spotify-tray: it should be able to find the running app and use it.

OK, I tried that command, it did change the WM_CLASS of the spotify window, confirmed when running `xprop | grep WM_CLASS` but when starting spority-tray it opens a new spotify window that has no WM_CLASS. I also noticed that when closing spotify there's a little change on the window icon to briefly show the spotify icon 1 second before closing.

The desktop file already has a WMClass set, it didn't changed on the update.
<details>
<summary>spotify.desktop</summary>

```
[Desktop Entry]
Type=Application
Name=Spotify
GenericName=Music Player
Icon=spotify-client
TryExec=spotify
Exec=spotify %U
Terminal=false
MimeType=x-scheme-handler/spotify;
Categories=Audio;Music;Player;AudioVideo;
StartupWMClass=spotify
```
</details>

I opened the deb files that are used by the AUR package and the only changes are on the binary.

I tried spotify on another DE, Awesome WM that I also have installed for usage with nvidia-xrun, and it also has no WM_Class, so no icon or spotify-tray either, then maybe there's a bug on X11 that surfaced by something weird on this new spotify binary.

Either way, this is not a bug on spotify-tray, it's either spotify or X11 fault, or even something else idk, all I can do is downgrade and wait. 

Thanks for the help debugging the issue, mate.
